pkg/image/registryclient/alternative-sources.go

```diff
@@ -0,0 +1,125 @@
+package registryclient
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/docker/distribution"
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	imagereference "github.com/openshift/library-go/pkg/image/reference"
+)
+
+// ImageContentSourcePolicy variables to look up alternative sources
+type AlternativeImageSources struct {
+	ICSPFile     string
+	ICSPList     []operatorv1alpha1.ImageContentSourcePolicy
+	ICSPClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)
+}
+
+type AlternativeSources interface {
+	PreferredImageSource(imageRef imagereference.DockerImageReference) (imagereference.DockerImageReference, distribution.Repository, distribution.ManifestService, error)
+}
+
+// AddICSPsFromCluster will lookup ICSPs from cluster. Unless explicit lookup is set to false, logs errors rather than returning errors.
+func (a *Context) AddICSPsFromCluster(explicit bool) error {
+	icspClient, err := a.ImageSources.ICSPClientFn()
+	if err != nil {
+		return err
+	}
+	if explicit && icspClient == nil {
+		return fmt.Errorf("flag --lookup-cluster-icsp was set to true, but no method was set to find ImageContentSourcePolicy objects in cluster")
+	}
+	if icspClient != nil {
```

Comment:
The above condition shouldn't matter on `explicit` and then this check is not needed.

Reply:
right, updated, fixed, and removed flag reference

Reply:
I need the check for icspClient != nil when !explicit, bc it will be nil whenever user is not connected to a cluster and in that case will get a `nil pointer dereference` when trying to list icsps